connectors/migrations/20241212_backfill_github_data_source_folders.ts

```diff
@@ -0,0 +1,236 @@
+import { makeScript } from "scripts/helpers";
+
+import { getGithubCodeDirectoryParentIds } from "@connectors/connectors/github/lib/hierarchy";
+import {
+  getCodeRootInternalId,
+  getDiscussionsInternalId,
+  getIssuesInternalId,
+  getMimeTypeFromGithubContentNodeType,
+  getRepositoryInternalId,
+} from "@connectors/connectors/github/lib/utils";
+import { dataSourceConfigFromConnector } from "@connectors/lib/api/data_source_config";
+import { concurrentExecutor } from "@connectors/lib/async_utils";
+import { upsertDataSourceFolder } from "@connectors/lib/data_sources";
+import {
+  GithubCodeDirectory,
+  GithubCodeFile,
+  GithubCodeRepository,
+  GithubDiscussion,
+  GithubIssue,
+} from "@connectors/lib/models/github";
+import { ConnectorModel } from "@connectors/resources/storage/models/connector_model";
+import type { DataSourceConfig } from "@connectors/types/data_source_config";
+
+async function createFolderNodes() {
+  const connectors = await ConnectorModel.findAll({
+    where: {
+      type: "github",
+    },
+  });
+
+  for (const connector of connectors) {
+    const dataSourceConfig = dataSourceConfigFromConnector(connector);
+    const repositories = await GithubCodeRepository.findAll({
+      where: {
+        connectorId: connector.id,
+      },
+    });
+
+    for (const repository of repositories) {
+      // upsert repository as folder
+      // Throws if error
+      const repoFolderId = await upsertRepositoryFolderNode(
+        repository,
+        dataSourceConfig
+      );
+
+      // Upsert Code folder if we have some (file or directory)
+      const shouldCreateCodeFolder = await repositoryContainsCode(repository);
+      if (shouldCreateCodeFolder) {
+        await upsertCodeFolderNode(
+          repository.repoId,
+          repoFolderId,
+          dataSourceConfig
+        );
+      }
+
+      const directories = await GithubCodeDirectory.findAll({
+        where: {
+          connectorId: connector.id,
+        },
+      });
+      // Upsert directories as folders, in chunks
+      await concurrentExecutor(
+        directories,
+        async (directory) => {
+          await upsertDirectoryFolderNode(
+            directory,
+            repository.id,
+            dataSourceConfig
+          );
+        },
+        { concurrency: 16 }
+      );
+    }
+
+    // Upsert issues folder if we have some
+    const repoWithIssues = await getUniqueRepositoryIdsWithIssues(connector.id);
+    await concurrentExecutor(
+      repoWithIssues,
+      async (repoId) => {
+        // The node id is the same as the repoId
+        await upsertIssueFolderNode(
+          repoId,
+          getRepositoryInternalId(repoId),
+          dataSourceConfig
+        );
+      },
+      { concurrency: 16 }
```

Comment:
nit: adding concurrency here doesn't amount to anything since there won't be that many repo anyway

Reply:
I am not taking any action at the issue level in this script but yep I would have clearly done it otherwise

Reply:
Oh my bad on this one, very clear 👍